### PR TITLE
Jinmen/settings

### PR DESCRIPTION
**Pull request by jinghanm**
## Description
Add settings e2e test

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-osconfig/blob/main/CONTRIBUTING.md).
- [x] All unit tests are passing.
- [x] I have merged the latest `main` branch prior to this PR submission.
- [x] I submitted this PR against the `main` branch.

### CODE REVIEW

File: src/tests/e2etest/SettingsTests.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+            Twin twinPatch = CreateTwinPatch(ComponentName, desiredSettings);
+
+            var reportedObject = new Settings();
+            if (UpdateTwinBlockUntilUpdate(twinPatch))
+            {
+                reportedObject = GetSettingsReportObject();
+                // Wait until the reported properties are updated
+                DateTime startTime = DateTime.Now;
+                while(!IsSameByJson(desiredSettings, reportedObject) && (DateTime.Now - startTime).TotalSeconds < twinTimeoutSeconds)
```
Comment (AhmedBM) [inline src/tests/e2etest/SettingsTests.cs:110]: not understanding why we have to keep doing `IsSameByJson`? Once we set the setting, the next reported interval should contain the value. So all needed here is to wait, until the next update after the patch was applied, none of this json comparing is needed

**Review by AhmedBM — CHANGES_REQUESTED** [review]

File: src/tests/e2etest/SettingsTests.cs (hunk 1 above)
Comment (jinghanm) [reply src/tests/e2etest/SettingsTests.cs]: Thanks Ahmed. I have made all necessary changes as you requested.

File: src/tests/e2etest/E2ETest.cs [hunk 2]
```diff
@@ -262,6 +262,13 @@ public static void AreEqualByJson(object expected, object actual)
             Assert.AreEqual(expectedJson, actualJson);
         }
 
+        public static bool IsSameByJson(object expected, object actual)
```
Comment (AhmedBM) [inline src/tests/e2etest/E2ETest.cs:265]: Unnecessary reference, not seeing any usage.

File: src/tests/e2etest/SettingsTests.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            {
+                Assert.Fail("Module is disconnected or is disabled");
+            }
+
+            Twin twinPatch = CreateTwinPatch(ComponentName, desiredSettings);
+
+            var reportedObject = new Settings();
+            if (UpdateTwinBlockUntilUpdate(twinPatch))
+            {
+                reportedObject = GetSettingsReportObject();
```
Comment (AhmedBM) [inline src/tests/e2etest/SettingsTests.cs:107]: but were not testing against any expected values here? How do we know this is the correct result?

**Review by AhmedBM — CHANGES_REQUESTED** [review]
This test just ensures the update is successful, but doesn't validate the expected state, we should make sure to also validate our expected state

File: src/tests/e2etest/SettingsTests.cs (hunk 3 above)
Comment (jinghanm) [reply src/tests/e2etest/SettingsTests.cs:107]: I used to have IsSameByJson(desiredSettings, reportedObject) in the while loop as testing against the expected values.

Comment (jinghanm) [reply src/tests/e2etest/SettingsTests.cs:94]: Add testing again the expected values.

File: src/tests/e2etest/E2ETest.cs (hunk 2 above)
Comment (jinghanm) [reply src/tests/e2etest/E2ETest.cs:265]: Thanks. Removed.

**Review by AhmedBM — APPROVED** [review]

**Pull request merged**